### PR TITLE
multi: add BitcoinCli support and MineBlock IR operation

### PR DESCRIPTION
**Pull request by NishantBansal2003**
Added a `BitcoinCli` struct that contains the connection information required to invoke bitcoin-cli against the regtest bitcoind started by a target, and added it to the program context. This will be useful for future IR funding flow support, where the IR executor may need a bitcoin-cli reference to mine blocks, broadcast tx etc. The executor can access this `BitcoinCli` reference through `ProgramContext`.

Added a `BitcoinCli` variable type and `LoadBitcoinCliFromContext` so IR programs can access the bitcoin-cli connection handle from the program context and make Bitcoin RPC calls.

Added a `MineBlock` IR operation that takes a `BitcoinCli` input and invokes `bitcoin-cli -generate 1` to mine a block on regtest. This will be required once we add the funding flow scenario, where the funding tx is broadcast and then mined into a block to confirm it on-chain, since peers can exchange the `channel_ready` message only after the funding tx is confirmed.

I also verified the working of `MineBlock` using the diff below:
<details>
<summary>smite diff</summary>

```diff
diff --git a/smite-ir/src/generators/open_channel.rs b/smite-ir/src/generators/open_channel.rs
index f58f1bf..bf94823 100644
--- a/smite-ir/src/generators/open_channel.rs
+++ b/smite-ir/src/generators/open_channel.rs
@@ -25,6 +25,7 @@ impl Generator for OpenChannelGenerator {
         let delayed_payment_basepoint = builder.generate_fresh(VariableType::Point, rng);
         let htlc_basepoint = builder.generate_fresh(VariableType::Point, rng);
         let first_per_commitment_point = builder.generate_fresh(VariableType::Point, rng);
+        let bitcoin_cli = builder.generate_fresh(VariableType::BitcoinCli, rng);
 
         // Channel parameters.
         let chain_hash = builder.pick_variable(VariableType::ChainHash, rng);
@@ -77,5 +78,7 @@ impl Generator for OpenChannelGenerator {
 
         // Receive accept_channel.
         builder.append(Operation::RecvAcceptChannel, &[]);
+
+        builder.append(Operation::MineBlock,&[bitcoin_cli]);
     }
 }
 ```

</details>

<details>
<summary>smite and bitcoind logs</summary>

```bash
INFO  [smite_scenarios::targets::bitcoind] Preserving data directory: /data
INFO  [smite_scenarios::targets::bitcoind] Starting bitcoind...
INFO  [smite_scenarios::targets::bitcoind] Waiting for bitcoind to be ready...
INFO  [smite_scenarios::targets::bitcoind] bitcoind is ready
INFO  [smite_scenarios::targets::ldk] Starting ldk-node-wrapper...
INFO  [smite_scenarios::targets::ldk] LDK identity pubkey: 03a74fcad0445473559e6a047f3b8b3b9ae758809fdd41ebe14e11b8c63342a1fc
INFO  [smite_scenarios::targets::ldk] ldk-node-wrapper is ready and synced
INFO  [smite_scenarios::targets::ldk] Both daemons are running, ready to fuzz
DEBUG [smite_scenarios::scenarios] Handshake complete, received target init
INFO  [smite::scenarios] Scenario initialized! Executing input...
INFO  [smite::runners] Reading input from "/input.bin"
DEBUG [smite_scenarios::scenarios::ir] [4.55µs] Executing IR program (27 instructions, 344 input bytes)
DEBUG [smite_scenarios::executor] [223.133µs] SendMessage: type Some(32), 356 bytes
DEBUG [smite_scenarios::executor] [236.478µs] RecvAcceptChannel: waiting
DEBUG [smite_scenarios::executor] [5.226548ms] RecvAcceptChannel: received
DEBUG [smite_scenarios::executor] [5.231948ms] MineBlock: mining a block
DEBUG [smite_scenarios::scenarios::ir] [9.390277ms] Program executed successfully
DEBUG [smite_scenarios::scenarios::ir] [9.589365ms] Target responded with pong
INFO  [smite::scenarios] Test case ran successfully!
DEBUG [smite::process] ldk-node-wrapper: dropping running process, attempting shutdown
DEBUG [smite::process] ldk-node-wrapper: sending SIGTERM to process group 53
DEBUG [smite::process] ldk-node-wrapper: exited with exit status: 0
DEBUG [smite::process] bitcoind: dropping running process, attempting shutdown
DEBUG [smite::process] bitcoind: sending SIGTERM to process group 7
DEBUG [smite::process] bitcoind: exited with exit status: 0

.......

2026-05-11T07:44:55Z UpdateTip: new best=754887284a582ef44b1875e9b9cfb1202812910f3bfb92ea6fca60b1a529910b height=101 version=0x20000000 log2_work=7.672425 tx=102 date='2026-05-11T07:45:13Z' progress=1.000000 cache=0.3MiB(101txo)
2026-05-11T07:44:55Z [default] AddToWallet ceabfb8fe0347dda0084f4a18752d1a15b7090a5ef45218c86c8e43826da8d71  new Confirmed (block=754887284a582ef44b1875e9b9cfb1202812910f3bfb92ea6fca60b1a529910b, height=101, index=0)
2026-05-11T07:44:55Z CreateNewBlock(): block weight: 888 txs: 0 fees: 0 sigops 400
2026-05-11T07:44:55Z Saw new header hash=29bf472f68bcc88a1a3e02c67331cfa4d2f956618777d1bbbcf281c5876d202b height=102
2026-05-11T07:44:55Z UpdateTip: new best=29bf472f68bcc88a1a3e02c67331cfa4d2f956618777d1bbbcf281c5876d202b height=102 version=0x20000000 log2_work=7.686501 tx=103 date='2026-05-11T07:45:13Z' progress=1.000000 cache=0.3MiB(102txo)
2026-05-11T07:44:55Z [default] AddToWallet 3325f6b56380693832efb8cacfaac961c60f0bbd983b2bf52006dad9d0b9c765  new Confirmed (block=29bf472f68bcc88a1a3e02c67331cfa4d2f956618777d1bbbcf281c5876d202b, height=102, index=0)
2026-05-11T07:44:55Z tor: Thread interrupt
```

</details>

### CODE REVIEW

File: smite-ir/src/operation.rs [hunk 1]
```diff
@@ -53,6 +53,8 @@ pub enum Operation {
     LoadTargetPubkeyFromContext,
     /// Load the chain hash from the program context.
     LoadChainHashFromContext,
+    /// Load the `bitcoin-cli` connection handle from the program context.
+    LoadBitcoinCliFromContext,
```
Comment (morehouse) [inline smite-ir/src/operation.rs]: I'm not sure we get any benefit from having a `BitcoinCli` variable type.  For the foreseeable future, I think there will only ever be one `bitcoind`, so we can simply use it directly in the executor.

File: smite-ir/src/operation.rs [hunk 2]
```diff
@@ -97,6 +99,9 @@ pub enum Operation {
     /// Receive and parse an `accept_channel` response.
     /// Produces an `AcceptChannel` compound variable.
     RecvAcceptChannel,
+    /// Mine a single block in Bitcoin network.
+    /// Input: `BitcoinCli`.
+    MineBlock,
```
Comment (morehouse) [inline smite-ir/src/operation.rs]: It *might* be useful to have a `u8` param for the number of blocks to mine, to reduce the number of RPC calls when doing channel funding.

We'd also want to test how slow `MineBlock(255)` is -- if it's really slow then we probably want to cap the number of blocks in the parameter mutator.

File: smite-scenarios/src/executor.rs [hunk 3]
```diff
@@ -999,6 +1028,67 @@ mod tests {
         assert!(matches!(err, ExecuteError::VoidVariable { index: 21 }));
     }
 
+    // MineBlock shells out to `bitcoin-cli -generate 1`, which only
+    // succeeds against a live regtest bitcoind. We don't run one in unit
+    // tests, so `mine_block` panics. The point of this test is not to mine
+    // a block but to verify the executor reaches `mine_block`.
+    #[test]
+    #[should_panic(expected = "bitcoin-cli -generate")]
+    fn execute_mine_blocks_invokes_cli() {
```
Comment (morehouse) [inline smite-scenarios/src/executor.rs]: I'm not sure about this test.  It seems to depend on `bitcoin-cli` not being present on the system, which we have no control over.

If we want to test this properly, we might need to mock the `BitcoinCli` like we do for the connection.

File: smite-scenarios/src/executor.rs [hunk 4]
```diff
@@ -999,6 +1028,67 @@ mod tests {
...
+        let _ = execute(
+            &program,
+            &sample_context(),
+            &mut conn,
+            std::time::Instant::now(),
+        );
+    }
+
+    #[test]
+    fn execute_mine_blocks_type_mismatch() {
```
Comment (morehouse) [inline smite-scenarios/src/executor.rs]: Nit on all test names: s/mine_blocks/mine_block

File: smite-scenarios/src/targets/bitcoind.rs [hunk 5]
```diff
@@ -127,30 +131,21 @@ pub fn start(config: &BitcoindConfig, data_dir: &Path) -> Result<ManagedProcess,
...
+) -> Result<(ManagedProcess, BitcoinCli), TargetError> {
     // Create wallet
-    let _ = Command::new("bitcoin-cli")
-        .arg("-regtest")
-        .arg(format!("-datadir={}", bitcoind_dir.display()))
-        .arg(format!("-rpcport={}", config.rpc_port))
-        .arg("-rpcuser=rpcuser")
-        .arg("-rpcpassword=rpcpass")
+    let _ = cli
+        .run()
```
Comment (morehouse) [inline smite-scenarios/src/targets/bitcoind.rs:137]: `bitcoind` is completely unused and `cli` is unmodified in this function.  Perhaps we should remove `bitcoind` and make `cli` a reference.

**Review by morehouse — CHANGES_REQUESTED** [review]
I think we should simplify by removing the new `BitcoinCli` variable type.  I don't think we actually need it.

**Comment by NishantBansal2003** [comment]
So, `minimum_depth` (in `accept_channel`) caps across LN implementations:
- LND:
    - Upper cap is 144: https://github.com/lightningnetwork/lnd/blob/cc65a20335f3206b3465e7e2ee002ec8c8226d12/funding/manager.go#L2133 (https://github.com/lightningnetwork/lnd/blob/cc65a20335f3206b3465e7e2ee002ec8c8226d12/chainntnfs/txnotifier.go#L16-L26)
    - Lower cap is 0: https://github.com/lightningnetwork/lnd/blob/cc65a20335f3206b3465e7e2ee002ec8c8226d12/funding/manager.go#L2153
- CLN:
    - Plugin-based, and I could not really find any upper cap (even BOLT does not specify any hard upper limit)
-  LDK
    - Upper cap (default) is 144: https://github.com/lightningdevkit/rust-lightning/blob/265e976c456fe1ecd7979050726f3f59a9b95a1e/lightning/src/util/config.rs#L406
    - Lower cap is 0: https://github.com/lightningdevkit/rust-lightning/blob/265e976c456fe1ecd7979050726f3f59a9b95a1e/lightning/src/ln/channel.rs#L4972
- Eclair:
    - Lower cap is 6 (0 for `option_zero_conf` channels): https://github.com/ACINQ/eclair/blob/8876ef184a2eea242ea6a07f117b9bc0ef6c791b/eclair-core/src/main/scala/fr/acinq/eclair/NodeParams.scala#L379 
    - I also could not really find any maximum cap here.

If we are dealing with outbound channels (where we do not control `minimum_depth`), mining 6-8 blocks is reasonable, since eclair has a default `minimum_depth` of 8 blocks.

If we are dealing with inbound channels later on (where we do control `minimum_depth`), I think 0-144 is a reasonable range to test.


Also, here is the timing for `MineBlocks` across some inputs. For values mostly > 50, we need to pass -t > 1000; otherwise, we start encountering tmouts:
```bash
DEBUG [smite_scenarios::executor] [16.324741ms] MineBlocks: mined 10 block(s) in 16.226165ms

DEBUG [smite_scenarios::executor] [85.38699ms] MineBlocks: mined 50 block(s) in 85.146694ms

DEBUG [smite_scenarios::executor] [157.109619ms] MineBlocks: mined 100 block(s) in 156.96939ms

DEBUG [smite_scenarios::executor] [252.175529ms] MineBlocks: mined 144 block(s) in 252.082276ms

DEBUG [smite_scenarios::executor] [273.198832ms] MineBlocks: mined 150 block(s) in 272.995616ms

DEBUG [smite_scenarios::executor] [359.333064ms] MineBlocks: mined 200 block(s) in 359.217774ms

DEBUG [smite_scenarios::executor] [468.346424ms] MineBlocks: mined 255 block(s) in 468.135552ms
```
I think having an upper cap of 150 for `MineBlocks` (and a lower cap of 1 to prevent invalid RPC calls to bitcoind) is the reasonable choice, since it widely exercises the upper bound cases for major LN implementations while also including a +6 buffer for additional edge case testing. Beyond that, I do not think mining more blocks provides much practical value

File: smite-ir/src/tests.rs [hunk 6]
```diff
@@ -878,6 +953,29 @@ fn param_mutator_changes_values() {
...
+
+    for _ in 0..100 {
+        mutator.mutate(&mut program, &mut rng);
+        // Make sure that MineBlocks contains a value within the clamped range of
+        // blocks to be mined.
+        if let Operation::MineBlocks(v) = program.instructions[0].operation {
+            assert!((1..=150).contains(&v));
+        }
+    }
+    assert_ne!(program, original);
```
Comment (morehouse) [inline smite-ir/src/tests.rs]: There's a non-negligible probability that `program == original here` (1/150).  Probably what we want is a success flag once `program != original`, or maybe a count of different programs.


```suggestion
    for _ in 0..100 {
        mutator.mutate(&mut program, &mut rng);
        // Make sure that MineBlocks contains a value within the clamped range of
        // blocks to be mined.
        let Operation::MineBlocks(num_blocks) = program.instructions[0].operation else {
            panic!("OperationParamMutator changed the operation type");
        }
        assert!((1..=150).contains(&num_blocks));
        if program != original {
            diff_count += 1;
        }
    }
    assert(diff_count > 90, "OperationParamMutator has an unexpected bias");
```

File: smite-ir/src/operation.rs [hunk 7]
```diff
@@ -97,6 +97,9 @@ pub enum Operation {
     /// Receive and parse an `accept_channel` response.
     /// Produces an `AcceptChannel` compound variable.
     RecvAcceptChannel,
+
+    /// Mines the given number of blocks on the Bitcoin network.
+    MineBlocks(u8),
```
Comment (morehouse) [inline smite-ir/src/operation.rs]:
```suggestion
    /// Mines the given number of blocks on the Bitcoin regtest network.
    MineBlocks(u8),
```

File: smite/src/bitcoin.rs [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        cmd.arg("-regtest")
+            .arg(format!("-datadir={}", self.bitcoind_dir.display()))
+            .arg(format!("-rpcport={}", self.rpc_port))
+            .arg("-rpcuser=rpcuser")
+            .arg("-rpcpassword=rpcpass");
+        cmd
+    }
+
+    /// Mines the given number of blocks.
+    #[allow(clippy::missing_panics_doc)]
```
Comment (morehouse) [inline smite/src/bitcoin.rs]: There should be a `Panics` doc section here.

File: smite-ir/src/mutators/operation_param.rs [hunk 9]
```diff
@@ -65,6 +65,13 @@ fn mutate_operation(op: &mut Operation, rng: &mut impl Rng) -> bool {
             mutate_channel_type(variant, rng);
             true
         }
+        Operation::MineBlocks(v) => {
+            // Ensure the number of mined blocks is always within a safe range.
+            // We cap it at 150 to avoid excessively long mining times, and
+            // ensure it is at least 1 to prevent invalid RPC calls to bitcoind.
+            *v = tweak_u8(*v, rng).clamp(1, 150);
```
Comment (morehouse) [inline smite-ir/src/mutators/operation_param.rs]: I'm thinking initially we should clamp this even further.  270ms at 150 blocks is a long time for a single operation, and I don't think we currently would get any benefit from mining more than ~10 blocks (AFAIK 8 blocks is the maximum default value for minimum_depth).

We can revisit later if we think the benefit of mining more blocks outweighs the risks.
```suggestion
            // Limit the number of mined blocks to keep execution times low.
            // Reference execution timings:
            // MineBlocks(10): 16ms
            // MineBlocks(100): 157ms
            // MineBlocks(200): 359ms
            // MineBlocks(255): 468ms
            *v = rng.random_range(1..=16);
```

File: smite-ir/src/tests.rs [hunk 10]
```diff
@@ -237,6 +242,62 @@ fn postcard_roundtrip() {
...
+fn displays_mine_blocks_program() {
+    let program = Program {
+        instructions: vec![Instruction {
+            operation: Operation::MineBlocks(6),
+            inputs: vec![],
+        }],
+    };
+    let text = program.to_string();
+    let lines: Vec<&str> = text.lines().collect();
+    assert_eq!(lines, vec!["MineBlocks(6)".to_string()],);
```
Comment (morehouse) [inline smite-ir/src/tests.rs]:
```suggestion
    assert_eq!(lines, vec!["MineBlocks(6)"]);
```

File: smite-scenarios/src/executor.rs [hunk 11]
```diff
@@ -23,6 +48,9 @@ pub struct ProgramContext {
     pub block_height: u32,
     /// Target's advertised feature bits from init message.
     pub target_features: Vec<u8>,
+    /// `bitcoin-cli` wrapper for interacting with the target's regtest
+    /// `bitcoind` instance.
+    pub bitcoin_cli: Rc<dyn BitcoinRpc>,
```
Comment (morehouse) [inline smite-scenarios/src/executor.rs]: The API becomes simpler if we just pass `&mut impl BitcoinRpc` to `execute` instead of trying to store it in `ProgramContext`.  If we follow the pattern used for the `Connection` trait, we get rid of all the `Rc`, `RefCell`, and `dyn BitcoinRpc` stuff.

File: smite-scenarios/src/executor.rs [hunk 12]
```diff
@@ -173,6 +201,18 @@ pub fn execute(
...
+
+            Operation::MineBlocks(v) => {
+                let mine_start = Instant::now();
+                context.bitcoin_cli.mine_blocks(*v);
+                log::debug!(
+                    "[{:?}] MineBlocks: mined {} block(s) in {:?}",
+                    start.elapsed(),
+                    v,
+                    mine_start.elapsed()
+                );
```
Comment (morehouse) [inline smite-scenarios/src/executor.rs]: Nit: we can infer the execution time using prepended time like the other match arms.


```suggestion
                context.bitcoin_cli.mine_blocks(*v);
                log::debug!(
                    "[{:?}] MineBlocks: mined {} block(s)",
                    start.elapsed(),
                    v,
                );
```

**Review by morehouse — CHANGES_REQUESTED** [review]

File: smite-scenarios/src/executor.rs (hunk 11 above)
Comment (NishantBansal2003) [reply smite-scenarios/src/executor.rs]: Yeah, I think that would be better

File: smite-scenarios/src/scenarios/ir.rs [hunk 13]
```diff
@@ -24,10 +26,11 @@ pub struct IrScenario<T: Target, S: SnapshotSetup<T>> {
 impl<T: Target, S: SnapshotSetup<T>> Scenario for IrScenario<T, S> {
     fn new(_args: &[String]) -> Result<Self, ScenarioError> {
         let target = T::start(T::Config::default())?;
-        let (conn, context) = S::setup(&target)?;
+        let (conn, bitcoin_cli, context) = S::setup(&target)?;
         Ok(Self {
             target,
             conn,
+            bitcoin_cli,
```
Comment (morehouse) [inline smite-scenarios/src/scenarios/ir.rs]: `bitcoin_cli` is already in `target`, so we shouldn't need it here too.

File: smite/src/bitcoin.rs [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+            .arg("-rpcpassword=rpcpass");
+        cmd
+    }
+
+    /// Mines the given number of blocks.
+    ///
+    /// # Panics
+    ///
+    /// If the `bitcoin-cli -generate` command fails to
+    /// execute or returns a non-success exit status.
```
Comment (morehouse) [inline smite/src/bitcoin.rs]: Nit: line wrap seems short here

File: smite-scenarios/src/scenarios/ir.rs (hunk 13 above)
Comment (NishantBansal2003) [reply smite-scenarios/src/scenarios/ir.rs]: Thanks, I got so blinded by conn and context that I completely missed that target was already there

**Review by morehouse — APPROVED** [review]
LGTM
